python/cudf/cudf/_lib/groupby.pyx

```diff
@@ -156,53 +150,55 @@ cdef class GroupBy:
                 c_agg_requests.push_back(
                     move(c_agg_request)
                 )
-
+                column_orders.append(i)
         if c_agg_requests.empty() and not allow_empty:
             raise DataError("All requested aggregations are unsupported.")
 
-        cdef pair[
-            unique_ptr[table],
-            vector[libcudf_groupby.aggregation_result]
-        ] c_result
-
         with nogil:
             c_result = move(
                 self.c_obj.get()[0].aggregate(
                     c_agg_requests
                 )
             )
 
-        grouped_keys, _ = data_from_unique_ptr(
-            move(c_result.first),
-            column_names=self.keys._column_names
+        grouped_keys = columns_from_unique_ptr(
+            move(c_result.first)
         )
 
-        result_data = ColumnAccessor(multiindex=True)
-        # Note: This loop relies on the included_aggregations dict being
-        # insertion ordered to map results to requested aggregations by index.
-        for i, col_name in enumerate(included_aggregations):
-            for j, agg_name in enumerate(included_aggregations[col_name]):
-                if callable(agg_name):
-                    agg_name = agg_name.__name__
-                result_data[(col_name, agg_name)] = (
-                    Column.from_unique_ptr(move(c_result.second[i].results[j]))
-                )
+        result_columns = [
+            [
+                Column.from_unique_ptr(
+                    move(c_result.second[i].results[j])
+                ) for j in range(c_result.second[i].results.size())
+            ] for i in range(c_result.second.size())
+        ]
+        result_columns_with_padding = [[]] * len(included_aggregations)
+        for i, cols in enumerate(result_columns):
+            result_columns_with_padding[column_orders[i]] = cols
```

Comment:
You're not observing any issues here because on line 177 you're overwriting the `column_orders[i]`th element, but just FYI this line is probably not doing what you expect. It's not create a list of separate lists, it's create a list of references that all point to the same list:
```
>>> x = [[]] *3
>>> x[0].append(1)
>>> x
[[1], [1], [1]]
```
The safe way to do this is with a list comprehension
```
>>> x = [[] for _ in range(3)]
>>> x[0].append(1)
>>> x
[[1], [], []]
```
I'm assuming that in this case if there are multiple columns that end up having no aggregations it doesn't break anything in the code if those empty lists happen to refer to the same underlying list. That said, perhaps a more idiomatic solution here would be something like 
```suggestion
        result_columns_with_padding = [
            cols if i in column_orders else []
            for i, cols in enumerate(result_columns)
        ]
```
In this case I would recommend constructing `column_orders` as a set instead of a list.

Reply:
Aren't we guaranteed that the order is identical except that some columns are just removed? As an example, say I have a DataFrame with columns 'a', 'b', and 'c', and then I request 2 aggregations for 'a', 0 for 'b', and 1 for 'c'. `column_orders` should be `[0, 0, 2]`, correct? And `result_columns` will look something like `[[first 'a' agg, second 'a' agg], [first 'c' agg]]`?
In particular, `column_orders` is always monotonically increasing up to duplicates. So then if we iterate through the results in the same order, we are guaranteed to put things in the right place as long as we skip indexes corresponding to input columns for which we performed no aggregations. Am I misunderstanding the logic?

Reply:
`column_orders` should be [0, 2]. One order per (input) column.

Reply:
Oh right, I made this mistake before too. It looks like it's inside the `for agg in aggs` loop because it is indented in the `if not c_agg_request.aggregations.empty():` block. But the absence of duplicates doesn't change the rest of my argument, does it?

Reply:
Closing the loop: the only thing missing from above is that we where to insert the libcudf result to the final column if we store `columns_order` as a `set`. But we DO know **when** and **what** to insert: when we iterate the result column and have a hit of position in `column_order`, we get the first column from the libcudf result. This can be implemented efficiently with a FIFO queue. For detail savvy friends: python list is rather inefficient if we were to use it as a queue (`O(N)` to pop from head, O(1) to pop from tail), so we used it as a stack and constructed the list in reverse.